### PR TITLE
EOS-9005:Modified cfs_detach and cfs_destroy_orphaned_file to use FH

### PR DESCRIPTION
**Pull request by Anilsindhi**
# Cortx-fs Change Summary

## Problem Statement

_[EOS-9005](https://jts.seagate.com/browse/EOS-9005):_
_EOS-NFS: Update cfs_detach, and cfs_destroy_orphaned_file_

## Problem Description

_With recent changes in FH we decided to supply a valid FH to all of the API in cortxfs wherever it is required, so that task was divided in further small task and as a part of that we have to modify cfs_detach and cfs_destroy_orphaned_file to use FH and optimize whatever was possible_

## Solution Overview

_As cfs_detach and cfs_destroy_orphaned_file were used by cortxfs API and FSAL code also so it was not directly possible to modify the signature of both of these API to pass FH as input parameter because we are not yet in a stage where we can pass a FH from FSAL code as it will have stale data. So with this limitation on both of these APIs I have created a temporary API where input will be FH and original API signature will remain same. So going forward the temporary API directly can be used by other APIs in cortxfs. Once all the caller of cfs_detach and cfs_destroy_orphaned_file has valid FH we can remove the temporary APIs_

## Unit Test Cases
_All UTs are passing_
_Able to create FS, mount FS and do basic I/O on that_
_Cthon is passing_
_Following steps are working as expected and output of this step has been added here as a testing summary_

1. sudo mount -o vers=4.0 127.0.0.1:/fs1 /tmp/mnt_dir/
2. cd /tmp/mnt_dir/
3. echo `date` > "file1"
4. mkdir dir1
5. for i in {1..5} ; do ln file1 dir1/link$i ; done
6. stat file1

## Note: 
_I have mounted /var/motr separately in order to avoid /var/motr is getting 100% full_

## Testing
<details>
  <summary>Click here to see the test execution output</summary>

```
# touch file1
[root@ssc-vm-0115: dir1 ]#
# ls
file1
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 0               Blocks: 0          IO Block: 1048576 regular empty file
Device: 2ch/44d Inode: 7           Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:40:35.362265817 -0600
Change: 2020-10-11 22:40:35.362267000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# mkdir dir1
[root@ssc-vm-0115: dir1 ]#
# ls
dir1  file1
[root@ssc-vm-0115: dir1 ]#
# echo `date` > file1
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:41:01.899546000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# for i in {1..5} ; do ln file1 dir1/link$i ; done
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 6
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:41:17.903080000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# rm dir1/link1
rm: remove regular file ‘dir1/link1’? y
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 5
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:41:49.540530000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# rm dir1/link2
rm: remove regular file ‘dir1/link2’? y
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 4
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:42:01.381457000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# ls
dir1  file1
[root@ssc-vm-0115: dir1 ]#
# cd dir1/
[root@ssc-vm-0115: dir1 ]#
# ls
link3  link4  link5
[root@ssc-vm-0115: dir1 ]#
# ls -lsi
total 2
7 1 -rw-r--r--. 4 root root 29 Oct 11 22:41 link3
7 1 -rw-r--r--. 4 root root 29 Oct 11 22:41 link4
7 1 -rw-r--r--. 4 root root 29 Oct 11 22:41 link5
[root@ssc-vm-0115: dir1 ]#
# cd ../
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 4
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:42:01.381457000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# rm dir1/link
link3  link4  link5
[root@ssc-vm-0115: dir1 ]#
# rm dir1/link3
rm: remove regular file ‘dir1/link3’? y
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 3
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:42:43.792214000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# rm dir1/link4
rm: remove regular file ‘dir1/link4’? y
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 2
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:42:53.528294000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# rm dir1/link5
rm: remove regular file ‘dir1/link5’? y
[root@ssc-vm-0115: dir1 ]#
# stat file1
  File: ‘file1’
  Size: 29              Blocks: 1          IO Block: 1048576 regular file
Device: 2ch/44d Inode: 7           Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:nfs_t:s0
Access: 2020-10-11 22:40:35.362265739 -0600
Modify: 2020-10-11 22:41:01.899546000 -0600
Change: 2020-10-11 22:43:01.791946000 -0600
 Birth: -
[root@ssc-vm-0115: dir1 ]#
# cat file1
Sun Oct 11 22:41:01 MDT 2020
```
  
</details>

## Checklist
- [x] **Compilation:** _This patch does not break compilation_
- [x] **Merge conflicts:** _This patch has been squashed and re-based, it can be merged using fast-forward merge_
- [x] **Code review:** _All discussions have been resolved_
- [x] **Sanity Testing:** _As mentioned in description all UTs and cthon is passing._
- [x] **Documentation:** _This patch, merge request, Jira ticket [EOS-9005](https://jts.seagate.com/browse/EOS-9005) have up to date description_

### CODE REVIEW

**Review by Shreya-18 — APPROVED** [review]
LGTM, +1 (no comments). Nice usage of temp functions.

File: src/cortxfs/cortxfs_ops.c [hunk 1]
```diff
@@ -828,48 +847,81 @@ int cfs_unlink(struct cfs_fs *cfs_fs, cfs_cred_t *cred, cfs_ino_t *dir,
...
+out:
+	if (parent_fh != NULL) {
+		cfs_fh_destroy_and_dump_stat(parent_fh);
+	}
+
+	if (child_fh != NULL) {
+		cfs_fh_destroy_and_dump_stat(child_fh);
+	}
+
+	log_trace("parent_ino=%llu name=%s child_ino=%llu rc=%d",
```
Comment (pratyush-seagate) [inline src/cortxfs/cortxfs_ops.c:924]: Looks like the code is not having any NULL check for the args  *parent_ino, name, *child_ino. If they ever be NULL, this trace will cause segfault.

File: src/cortxfs/cortxfs_ops.c [hunk 2]
```diff
@@ -550,15 +552,32 @@ int cfs_destroy_orphaned_file(struct cfs_fs *cfs_fs,
...
+	 * API when caller pass the valid FH instead of inode number
+	 */
+	RC_WRAP_LABEL(rc, out, cfs_fh_from_ino, cfs_fs, ino, &fh);
+	RC_WRAP_LABEL(rc, out, cfs_destroy_orphaned_file2, fh);
+
+out:
+	if (fh != NULL) {
+		cfs_fh_destroy(fh);
+	}
+	log_trace("inode=%llu rc=%d", *ino, rc);
```
Comment (pratyush-seagate) [inline src/cortxfs/cortxfs_ops.c:580]: Do we need to check for NULL before accessing *ino?

File: src/cortxfs/cortxfs_ops.c (hunk 1 above)
Comment (Anilsindhi) [reply src/cortxfs/cortxfs_ops.c:924]: `cfs_fh_from_ino` is doing this check and even caller of this APIs are checking that hence no check here.

File: src/cortxfs/cortxfs_ops.c (hunk 2 above)
Comment (Anilsindhi) [reply src/cortxfs/cortxfs_ops.c:580]: `cfs_fh_from_fh` is check here and caller of this APIs are also making sure hence no check. Also here using `ino` we are deriving `fh` and that is also validated in `cfs_fh_from_inode` before returning here. So I guess no validation is required.

**Review by s-arya — APPROVED** [review]
LGTM.

**Review by pratyush-seagate — APPROVED** [review]
LGTM, +1

**Pull request merged**